src/proxy/inbound.rs

```diff
@@ -606,51 +485,126 @@ impl Inbound {
+
+                let endpoint_port = svc
+                    .endpoints
+                    .get(&local_workload.uid)
+                    .and_then(|ep| ep.port.get(service_port));
+                // If we can get the port from the endpoint, that is ideal. But we may not, which is fine
+                // if the service has a number target port (rather than named).
+                let port = if let Some(&ep_port) = endpoint_port {
+                    ep_port
+                } else {
+                    let service_target_port =
+                        svc.ports.get(service_port).copied().unwrap_or_default();
+                    if service_target_port == 0 {
+                        return Err(Error::NoPortForServices(
+                            hostname.to_string(),
+                            *service_port,
+                        ));
+                    }
+                    service_target_port
+                };
+                (SocketAddr::new(target_ip, port), vec![svc])
+            }
+            HboneAddress::SocketAddr(hbone_addr) => (
+                SocketAddr::new(target_ip, hbone_addr.port()),
+                state.get_services_by_workload(local_workload),
+            ),
+        };
+
+        // Check for illegal calls now that we have resolved to the final destination.
+        // We need to do this here, rather than `validate_destination`, since the former doesn't
+        // have access to the resolved service port.
+        if cfg.illegal_ports.contains(&dest.port()) {
+            return Err(Error::SelfCall);
+        }
 
         // Application tunnel may override the port.
-        let (upstream_addr, inbound_protocol) = match local_workload.application_tunnel.clone() {
+        let (target, tunnel) = match local_workload.application_tunnel.clone() {
             Some(workload::ApplicationTunnel { port, protocol }) => {
                 // We may need to override the target port. For instance, we may send all PROXY
                 // traffic over a dedicated port like 15088.
-                let new_target =
-                    SocketAddr::new(upstream_addr.ip(), port.unwrap_or(upstream_addr.port()));
+                let new_target = SocketAddr::new(dest.ip(), port.unwrap_or(dest.port()));
                 // Note: the logic to decide which destination address to set inside the PROXY headers
                 // is handled outside of this call. This just determines that location we actually send the
-                // connection to
-                (new_target, protocol)
+                // connection to.
+
+                // Which address we will send in the tunnel
+                let tunnel_target = match hbone_addr {
+                    HboneAddress::SvcHostname(h, port) => {
+                        // PROXY cannot currently send to hostnames, so we will need to select an IP to
+                        // use instead
+                        let vip = services[0]
```

Comment:
This could panic if we happen to have no services, yeah? I think we want to use first() instead.

Reply:
Sounds good. I would prefer the expect personally.